python/BioSimSpace/Align/_merge.py

```diff
@@ -142,17 +142,9 @@ def merge(
 
     # Validate the region of interest.
     if roi is not None:
-        if not isinstance(roi, list):
-            raise TypeError("'roi' must be of type 'list'.")
-
-        for mol in [molecule0, molecule1]:
-            for idx in roi:
-                if not isinstance(idx, int):
-                    raise TypeError("'roi' must be a list of integers.")
-                if idx < 0 or idx > (mol.nResidues() - 1):
-                    raise ValueError(
-                        f"Residue index {idx} is out of range! The molecule contains {mol.nResidues()} residues."
-                    )
+        from ..Align import _validate_roi
```

Comment:
I'm actually confused why this works, since `_validate_roi` isn't exposed at the top level in the `Align` submodule, i.e. it's just a private function within the `BioSimSpace.Align._align` module. I would have thought that you want:

```python
from ._align import _validate_roi
```

Indeed, if I do:

```python
In [1]: import BioSimSpace as BSS


In [2]: help(BSS.Align._validate_roi)
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:1                                                                                    │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: module 'BioSimSpace.Align' has no attribute '_validate_roi'

In [3]: help(BSS.Align._align._validate_roi)
Help on function _validate_roi in module BioSimSpace.Align._align:

_validate_roi(molecules, roi)
    Internal function to validate that a region of interest (ROI) is a list
    of integers that are within the range of the molecule.

    Parameters
    ----------

    molecules : list
        Consits of a list of :class:`Molecule <BioSimSpace._SireWrappers.Molecule>`.

    roi : list
        The region of interest.
        Consists of a list of ROI residue indices.
```

which is what I would expect.

Reply:
If I run the test after adding the following patch:

```
diff --git a/python/BioSimSpace/Align/_merge.py b/python/BioSimSpace/Align/_merge.py
index 936dc808..c8a9bfa1 100644
--- a/python/BioSimSpace/Align/_merge.py
+++ b/python/BioSimSpace/Align/_merge.py
@@ -141,6 +141,7 @@ def merge(
                 )

     # Validate the region of interest.
+    print("merge", roi)
     if roi is not None:
         from ..Align import _validate_roi

diff --git a/tests/Align/test_align.py b/tests/Align/test_align.py
index a9a9d55c..dfcc6bde 100644
--- a/tests/Align/test_align.py
+++ b/tests/Align/test_align.py
@@ -669,6 +669,9 @@ def test_roi_merge(protein_inputs):
     p0 = BSS.Parameters.ff14SB(p0).getMolecule()
     p1 = BSS.Parameters.ff14SB(p1).getMolecule()

+    print()
+    print("test_roi", roi)
+
     aligned_p0 = BSS.Align.rmsdAlign(p0, p1, roi=roi)
     merged = BSS.Align.merge(aligned_p0, p1, protein_mapping)
     merged_system = merged.toSystem()
```

then I see the following in the output:

```
test_roi [2]
merge None
PASSED
tests/Align/test_align.py::test_roi_merge[double_neighbour_mutant] Downloading from 'https://biosimspace.openbiosim.org/m/double_neighbour_mutant_mut_peptide.pdb'...
Unzipping '/tmp/tmptdso2t4c/double_neighbour_mutant_mut_peptide.pdb.bz2'...
Downloading from 'https://biosimspace.openbiosim.org/m/double_neighbour_mutant_wt_peptide.pdb'...
Unzipping '/tmp/tmpj86a7v_q/double_neighbour_mutant_wt_peptide.pdb.bz2'...

test_roi [2, 3]
merge None
PASSED
```

So the `roi` is correct in the unit test, but is `None` within the merge function itself, which is because the `kwarg` isn't being used in the `merge` call:

```python
    aligned_p0 = BSS.Align.rmsdAlign(p0, p1, roi=roi)
    merged = BSS.Align.merge(aligned_p0, p1, protein_mapping)
    merged_system = merged.toSystem()
```

Reply:
I think the following fixes things:

```
diff --git a/python/BioSimSpace/Align/_merge.py b/python/BioSimSpace/Align/_merge.py
index 936dc808..ba8b6129 100644
--- a/python/BioSimSpace/Align/_merge.py
+++ b/python/BioSimSpace/Align/_merge.py
@@ -142,9 +142,9 @@ def merge(

     # Validate the region of interest.
     if roi is not None:
-        from ..Align import _validate_roi
+        from ._align import _validate_roi

-        _validate_roi(roi, molecule0, molecule1)
+        _validate_roi([molecule0, molecule1], roi)

     # Set 'allow_ring_breaking' and 'allow_ring_size_change' to true if the
     # user has requested to 'force' the merge, i.e. the 'force' argument
diff --git a/tests/Align/test_align.py b/tests/Align/test_align.py
index a9a9d55c..fa6432ed 100644
--- a/tests/Align/test_align.py
+++ b/tests/Align/test_align.py
@@ -670,6 +670,6 @@ def test_roi_merge(protein_inputs):
     p1 = BSS.Parameters.ff14SB(p1).getMolecule()

     aligned_p0 = BSS.Align.rmsdAlign(p0, p1, roi=roi)
-    merged = BSS.Align.merge(aligned_p0, p1, protein_mapping)
+    merged = BSS.Align.merge(aligned_p0, p1, mapping=protein_mapping, roi=roi)
     merged_system = merged.toSystem()
     assert merged_system.nPerturbableMolecules() == 1
```

Reply:
Ah, that makes sense. It's supposed to have `roi` keyword there, but I forgot to add it. I guess for this small peptide you can use it without it, but if I had a test with a much bigger protein then it would take hours to finish.

Reply:
If I expose `_validate_roi` in `Align.__init__.py` then I can properly import it in the merge function:
```bash
diff --git a/python/BioSimSpace/Align/__init__.py b/python/BioSimSpace/Align/__init__.py
index 47f97664..c28ed22e 100644
--- a/python/BioSimSpace/Align/__init__.py
+++ b/python/BioSimSpace/Align/__init__.py
@@ -41,3 +41,4 @@ Functions
 """

 from ._align import *
+from ._align import _validate_roi
diff --git a/python/BioSimSpace/Align/_merge.py b/python/BioSimSpace/Align/_merge.py
index 936dc808..8850352b 100644
--- a/python/BioSimSpace/Align/_merge.py
+++ b/python/BioSimSpace/Align/_merge.py
@@ -143,7 +143,6 @@ def merge(
     # Validate the region of interest.
     if roi is not None:
         from ..Align import _validate_roi
-
         _validate_roi(roi, molecule0, molecule1)

     # Set 'allow_ring_breaking' and 'allow_ring_size_change' to true if the
```

Reply:
Okay, nevermind. It's because I was using `from ..Align import _validate_roi` rather than `from ._align import _validate_roi` like how you're using.

Reply:
Hopefully that should be to go now!

Reply:
Yes, don't expose it at the `__init__.py` level since this is internal functionality only, so the user shouldn't see it.

Reply:
Looks like it just needs to be blackened, otherwise all should be okay.

Reply:
Done!